File: examples/mobileclip2/cpp/mobileclip2.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <algorithm>
#include <iostream>  // NOLINT(build/include_order)
#include <iomanip>   // NOLINT(build/include_order)
#include <memory>    // NOLINT(build/include_order)
#include <sstream>   // NOLINT(build/include_order)
#include <string>    // NOLINT(build/include_order)
#include <variant>   // NOLINT(build/include_order)
#include <vector>    // NOLINT(build/include_order)

#include "vision_service.h"  // NOLINT(build/include_order)

static std::vector<std::string> split_csv(const std::string& csv) {
    std::vector<std::string> out;
    std::stringstream ss(csv);
    std::string item;
    while (std::getline(ss, item, ',')) {
        const auto start = item.find_first_not_of(" \t");
        const auto end = item.find_last_not_of(" \t");
        if (start == std::string::npos) {
            continue;
        }
        out.push_back(item.substr(start, end - start + 1));
    }
    return out;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: " << argv[0] << " <config_yaml> [options]\n"
            << "  Options: --image <path> --text \"label1,label2\" --model-path <path>"
            << std::endl;
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string text_csv = "a photo of a dog,a photo of a cat,a photo of a car";
    std::string model_path_override;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--text" && i + 1 < argc) {
            text_csv = argv[++i];
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    std::unique_ptr<VisionService> service =
        VisionService::Create(config_path, model_path_override, true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (image_path.empty()) {
        image_path = service->GetDefaultImage();
    }
    if (image_path.empty()) {
        std::cerr << "Error: No image path. Use --image or set test_image in config." << std::endl;
        return 1;
    }

    VisionServiceResponse response;
    if (service->Infer(image_path, &response) != VISION_SERVICE_OK) {
        std::cerr << "Error: " << service->LastError() << std::endl;
        return 1;
    }
    if (response.results.empty()) {
        std::cerr << "Error: No image embedding returned." << std::endl;
        return 1;
    }
    const vision::Embedding* image_emb = std::get_if<vision::Embedding>(&response.results[0]);
    if (image_emb == nullptr) {
        std::cerr << "Error: Model did not return an embedding." << std::endl;
        return 1;
    }

    const std::vector<std::string> labels = split_csv(text_csv);
    if (labels.empty()) {
        std::cerr << "Error: No text labels provided." << std::endl;
        return 1;
    }

    std::cout << "Image: " << image_path << std::endl;
    float best_score = -1.0f;
    std::string best_label;
    for (const std::string& label : labels) {
        std::vector<float> text_emb;
        if (service->EncodeText(label, &text_emb) != VISION_SERVICE_OK) {
            std::cerr << "Error encoding text: " << service->LastError() << std::endl;
            return 1;
        }
        const float score = VisionService::EmbeddingSimilarity(image_emb->embedding, text_emb);
        std::cout << "  " << label << " : " << std::fixed << std::setprecision(4) << score
            << std::endl;
        if (score > best_score) {
            best_score = score;
            best_label = label;
        }
    }
    std::cout << "Best match: " << best_label << " (" << best_score << ")" << std::endl;
    return 0;
}

File: examples/mobileclip2/python/mobileclip2.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
# SPDX-License-Identifier: Apache-2.0

import argparse
from pathlib import Path

import cv2
import yaml

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def parse_args():
    parser = argparse.ArgumentParser(description="MobileCLIP2 image-text embedding example")
    parser.add_argument("--config", type=str, default=None)
    parser.add_argument("--image", type=str, default=None)
    parser.add_argument(
        "--text",
        type=str,
        default="a photo of a dog,a photo of a cat,a photo of a car",
    )
    parser.add_argument("--model-path", type=str, default=None)
    return parser.parse_args()


def main():
    args = parse_args()
    default_config = Path(__file__).parent.parent / "config" / "mobileclip2.yaml"
    config_path = Path(args.config) if args.config else default_config
    project_root = Path(__file__).parent.parent.parent.parent

    config = {}
    if config_path.exists():
        with open(config_path, "r", encoding="utf-8") as f:
            config = yaml.safe_load(f) or {}

    image_path = args.image or config.get("test_image")
    if not image_path:
        raise ValueError("Provide --image or set test_image in config")
    image_path = resolve_path(image_path, project_root)

    model_path_override = ""
    if args.model_path:
        p = Path(args.model_path).expanduser()
        model_path_override = str(p if p.is_absolute() else (project_root / p).resolve())

    svc = VisionServiceNative.create(str(config_path), model_path_override=model_path_override)

    image = cv2.imread(str(image_path))
    if image is None:
        raise FileNotFoundError(f"Failed to load image: {image_path}")

    st, image_emb = svc.infer_embedding(image)
    if st != VisionServiceStatus.OK:
        raise RuntimeError(svc.last_error())

    labels = [t.strip() for t in args.text.split(",") if t.strip()]
    best_label = ""
    best_score = -1.0
    print(f"Image: {image_path}")
    for label in labels:
        st, text_emb = svc.encode_text(label)
        if st != VisionServiceStatus.OK:
            raise RuntimeError(svc.last_error())
        score = VisionServiceNative.embedding_similarity(image_emb, text_emb)
        print(f"  {label} : {score:.4f}")
        if score > best_score:
            best_score = score
            best_label = label
    print(f"Best match: {best_label} ({best_score:.4f})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
